src/lang/builtins_source.ml

```diff
@@ -139,6 +139,23 @@ let () =
       s#on_track (fun m -> ignore (Lang.apply f [("", Lang.metadata m)]));
       Lang.unit)
 
+let () =
+  add_builtin "source.on_leave" ~cat:Sys
+    [
+      ("", Lang.source_t (Lang.univ_t ()), None, None);
+      ("", Lang.fun_t [] Lang.unit_t, None, None);
+    ]
+    Lang.unit_t
+    ~descr:
+      "Register a function to be called when source is not used anymore by \
+       another souce."
+    (fun p ->
+      let s = Lang.to_source (Lang.assoc "" 1 p) in
+      let f = Lang.assoc "" 2 p in
+      let wrap_f () = ignore (Lang.apply f []) in
+      s#on_leave wrap_f;
+      Lang.unit)
```

Comment:
Do we need this?

Reply:
Yes in order to have decent types for functions in scripts.